### PR TITLE
[dapp-kit] create new wallet provider for managing wallet connection state

### PR DESCRIPTION
**Pull request by williamrobertson13**
## Description 

This PR is the first part in a series of work to add wallet connection management functionality to `dapp-kit`. The general plan is that we are going to use native state management in React combined with `react-query` to expose a new series of hooks that will replace `wallet-kit`. 

For this PR specifically, we're scaffolding out the initial context, state management, and new hook called `useWallet` that exposes some of the wallet connection state such as accounts and wallets. To answer a few potential questions on this one:

Why are we using a reducer for state management vs. normal `useState`, zustand, etc.?
-> Most/all of the wallet state is intertwined enough that it makes sense to use a reducer over `useState`. Since we're going for simplicity with the opinionated React architecture, we're going to roll with this for now over something like Zustand.

Why are we adding a `connectionStatus` state instead of relying on mutation state from `useConnectWallet`?
-> Mutation state can't be shared in react-query v4, so we need to manage this state manually (https://github.com/TanStack/query/issues/2304)

Why no tests?
-> These will be important to have, but setting up useful tests given the nature of this functionality is a little tricky. Depending on the resourcing for this, I'll work on some combination of core logic tests using a mock wallet, some dApp we spin up for E2E tests, or something along those lines.

## Test Plan
- I have an example Vite application that will be added in a later PR  that I've been testing with 
- CI

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: sdk/dapp-kit/src/index.ts [hunk 1]
```diff
@@ -3,8 +3,10 @@
 
 export * from './hooks/useSuiClient.js';
 export * from './components/SuiClientProvider.js';
+export * from './components/WalletProvider.js';
```
Comment (williamrobertson13) [inline sdk/dapp-kit/src/index.ts:6]: We'll likely just expose a `DAppKitProvider` hook that wraps `SuiClientProvider` and `WalletProvider`, but we'll do that in a different PR

File: sdk/dapp-kit/src/utils/storageAdapters.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
+// Copyright (c) Mysten Labs, Inc.
```
Comment (williamrobertson13) [inline sdk/dapp-kit/src/utils/storageAdapters.ts:1]: This is more-or-less copy + pasted from `wallet-kit`

File: sdk/dapp-kit/src/utils/walletUtils.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+// Copyright (c) Mysten Labs, Inc.
```
Comment (williamrobertson13) [inline sdk/dapp-kit/src/utils/walletUtils.ts:1]: This is also more-or-less copy + pasted from `wallet-kit`

File: sdk/dapp-kit/src/hooks/wallet/useWallet.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { useWalletContext } from '../../components/WalletProvider.js';
+import { useMemo } from 'react';
+
+/**
+ * Hook for retrieving wallet and account information.
+ */
+export function useWallet() {
```
Comment (Jordan-Mysten) [inline sdk/dapp-kit/src/hooks/wallet/useWallet.ts:10]: We talked about making this a bunch of hooks instead of one I thought

Comment (Jordan-Mysten) [reply sdk/dapp-kit/src/hooks/wallet/useWallet.ts:10]: Hmm, I actually think that this might be okay to have multiple hooks for each of these? I'm not 100% sure, and we can iterate on this in the future, but I think more often than not you're just grabbing the current account anyway, so having a bag of functionality that comes along for the ride isn't the most ideal.

Splitting hooks up _could_ improve performance if these each have their own provider, or if we implement some approach using selectors (maybe using zustand or jotai or whatever), but in the current approach yeah there's no perf benefit.

Comment (Jordan-Mysten) [reply sdk/dapp-kit/src/hooks/wallet/useWallet.ts:10]: But for now we can leave this as-is and iterate in the future

**Review by Jordan-Mysten — APPROVED** [review]

File: sdk/dapp-kit/src/hooks/wallet/useWallet.ts (hunk 4 above)
Comment (williamrobertson13) [reply sdk/dapp-kit/src/hooks/wallet/useWallet.ts:10]: Okay yeah, that makes sense and +1 to your point about the potential improvements to prevent re-renders. Let me get the base of all this fleshed out and we can iterate like you said!
